Reflect protocol v6 changes in schema JSON output

Recent changes in https://github.com/hashicorp/terraform/pull/27699 and https://github.com/hashicorp/terraform/pull/27793 introduced some new fields for an attribute, which should probably be reflected somehow in the schema JSON output produced by `terraform providers schema -json`.

The SDK doesn't seem to support v6 yet, so this _currently_ has no real impact on end-users AFAICT, but when I tried adding some of the new fields to the builtin provider I saw a crash.



### Terraform Version

Modified version of `e5538693ced0dc42bb90a3919875cb8bbd625736` (latest revision)

### Modified schema

https://github.com/hashicorp/terraform/blob/59b71de2befadf712a81d49b5e51226660ecc173/builtin/providers/terraform/data_source_state.go#L59-L76

(feel free to checkout `json-schema-export-crash` branch)

### Terraform Configuration Files

```terraform
resource "terraform_remote_state" "test" {

}
```

### Crash Output

```
2021-03-11T12:23:36.083Z [INFO]  Terraform version: 0.15.0 dev
2021-03-11T12:23:36.083Z [INFO]  Go runtime version: go1.16
2021-03-11T12:23:36.083Z [INFO]  CLI args: []string{"/Users/radeksimko/gopath/bin/terraform", "providers", "schema", "-json"}
2021-03-11T12:23:36.083Z [TRACE] Stdout is a terminal of width 182
2021-03-11T12:23:36.083Z [TRACE] Stderr is a terminal of width 182
2021-03-11T12:23:36.083Z [TRACE] Stdin is a terminal
2021-03-11T12:23:36.083Z [DEBUG] Attempting to open CLI config file: /Users/radeksimko/.terraformrc
2021-03-11T12:23:36.083Z [INFO]  Loading CLI configuration from /Users/radeksimko/.terraformrc
2021-03-11T12:23:36.084Z [DEBUG] checking for credentials in "/Users/radeksimko/.terraform.d/plugins"
2021-03-11T12:23:36.084Z [DEBUG] ignoring non-existing provider search directory terraform.d/plugins
2021-03-11T12:23:36.084Z [DEBUG] will search for provider plugins in /Users/radeksimko/.terraform.d/plugins
2021-03-11T12:23:36.084Z [DEBUG] ignoring non-existing provider search directory /Users/radeksimko/Library/Application Support/io.terraform/plugins
2021-03-11T12:23:36.084Z [DEBUG] ignoring non-existing provider search directory /Library/Application Support/io.terraform/plugins
2021-03-11T12:23:36.084Z [INFO]  CLI command args: []string{"providers", "schema", "-json"}
2021-03-11T12:23:36.085Z [TRACE] Meta.Backend: no config given or present on disk, so returning nil config
2021-03-11T12:23:36.085Z [TRACE] Meta.Backend: backend has not previously been initialized in this working directory
2021-03-11T12:23:36.085Z [DEBUG] New state was assigned lineage "c9080e6c-8149-656f-2396-72941fe0038b"
2021-03-11T12:23:36.085Z [TRACE] Meta.Backend: using default local state only (no backend configuration, and no existing initialized backend)
2021-03-11T12:23:36.085Z [TRACE] Meta.Backend: instantiated backend of type <nil>
2021-03-11T12:23:36.086Z [DEBUG] checking for provisioner in "."
2021-03-11T12:23:36.087Z [DEBUG] checking for provisioner in "/Users/radeksimko/gopath/bin"
2021-03-11T12:23:36.087Z [DEBUG] checking for provisioner in "/Users/radeksimko/.terraform.d/plugins"
2021-03-11T12:23:36.087Z [INFO]  Failed to read plugin lock file .terraform/plugins/darwin_amd64/lock.json: open .terraform/plugins/darwin_amd64/lock.json: no such file or directory
2021-03-11T12:23:36.087Z [TRACE] Meta.Backend: backend <nil> does not support operations, so wrapping it in a local backend
2021-03-11T12:23:36.087Z [TRACE] backend/local: requesting state manager for workspace "default"
2021-03-11T12:23:36.087Z [TRACE] backend/local: state manager for workspace "default" will:
 - read initial snapshot from terraform.tfstate
 - write new snapshots to terraform.tfstate
 - create any backup at terraform.tfstate.backup
2021-03-11T12:23:36.087Z [TRACE] backend/local: requesting state lock for workspace "default"
2021-03-11T12:23:36.087Z [TRACE] backend/local: reading remote state for workspace "default"
2021-03-11T12:23:36.087Z [TRACE] statemgr.Filesystem: reading initial snapshot from terraform.tfstate
2021-03-11T12:23:36.087Z [TRACE] statemgr.Filesystem: snapshot file has nil snapshot, but that's okay
2021-03-11T12:23:36.087Z [TRACE] statemgr.Filesystem: read nil snapshot
2021-03-11T12:23:36.087Z [DEBUG] backend/local: skipping refresh of managed resources
2021-03-11T12:23:36.087Z [TRACE] backend/local: retrieving local state snapshot for workspace "default"
2021-03-11T12:23:36.087Z [TRACE] backend/local: building context for current working directory
2021-03-11T12:23:36.087Z [TRACE] terraform.NewContext: starting
2021-03-11T12:23:36.087Z [TRACE] terraform.NewContext: loading provider schemas
2021-03-11T12:23:36.088Z [TRACE] LoadSchemas: retrieving schema for provider type "terraform.io/builtin/terraform"
2021-03-11T12:23:36.088Z [TRACE] terraform.NewContext: complete
2021-03-11T12:23:36.088Z [TRACE] backend/local: finished building terraform.Context

panic: unknown type implementation

goroutine 1 [running]:
github.com/zclconf/go-cty/cty.Type.MarshalJSON(0x0, 0x0, 0xd9c9d55d, 0x2fcb13150d2aa4a, 0xffffffffffffffff, 0x0, 0x0)
	/Users/radeksimko/gopath/pkg/mod/github.com/zclconf/go-cty@v1.8.0/cty/json.go:87 +0xd65
github.com/hashicorp/terraform/command/jsonprovider.marshalAttribute(0xc0005ab1c0, 0xc00036c330)
	/Users/radeksimko/gopath/src/github.com/hashicorp/terraform/command/jsonprovider/attribute.go:32 +0x3b
github.com/hashicorp/terraform/command/jsonprovider.marshalBlock(0xc00036c000, 0xc0004adc60)
	/Users/radeksimko/gopath/src/github.com/hashicorp/terraform/command/jsonprovider/block.go:63 +0x1fe
github.com/hashicorp/terraform/command/jsonprovider.marshalSchema(...)
	/Users/radeksimko/gopath/src/github.com/hashicorp/terraform/command/jsonprovider/schema.go:20
github.com/hashicorp/terraform/command/jsonprovider.marshalSchemas(0xc00036c060, 0xc00036c090, 0xc00036c2d0)
	/Users/radeksimko/gopath/src/github.com/hashicorp/terraform/command/jsonprovider/schema.go:31 +0x127
github.com/hashicorp/terraform/command/jsonprovider.marshalProvider(0xc00036c0c0, 0x339dc5b)
	/Users/radeksimko/gopath/src/github.com/hashicorp/terraform/command/jsonprovider/provider.go:62 +0xf0
github.com/hashicorp/terraform/command/jsonprovider.Marshal(0xc0004adc10, 0x0, 0x0, 0xc0005eb800, 0x1, 0x1)
	/Users/radeksimko/gopath/src/github.com/hashicorp/terraform/command/jsonprovider/provider.go:38 +0x305
github.com/hashicorp/terraform/command.(*ProvidersSchemaCommand).Run(0xc0001d2fc0, 0xc00004e0b0, 0x1, 0x1, 0xc0004ad170)
	/Users/radeksimko/gopath/src/github.com/hashicorp/terraform/command/providers_schema.go:100 +0x889
github.com/mitchellh/cli.(*CLI).Run(0xc000399180, 0xc0002fbb00, 0x33a6fb9, 0x10)
	/Users/radeksimko/gopath/pkg/mod/github.com/mitchellh/cli@v1.1.2/cli.go:262 +0x41a
main.wrappedMain(0x0)
	/Users/radeksimko/gopath/src/github.com/hashicorp/terraform/main.go:379 +0x10e8
main.realMain(0x0)
	/Users/radeksimko/gopath/src/github.com/hashicorp/terraform/main.go:115 +0x554
main.main()
	/Users/radeksimko/gopath/src/github.com/hashicorp/terraform/main.go:58 +0x25

```

### Expected Behavior

Output JSON schema

### Actual Behavior

Crash

### Steps to Reproduce

1. `terraform init`
2. `terraform providers schema -json`

--- a/command/jsonprovider/attribute.go
+++ b/command/jsonprovider/attribute.go
@@ -4,17 +4,26 @@ import (
 	"encoding/json"
 
 	"github.com/hashicorp/terraform/configs/configschema"
+	"github.com/zclconf/go-cty/cty"
 )
 
 type attribute struct {
-	AttributeType   json.RawMessage `json:"type,omitempty"`
-	Description     string          `json:"description,omitempty"`
-	DescriptionKind string          `json:"description_kind,omitempty"`
-	Deprecated      bool            `json:"deprecated,omitempty"`
-	Required        bool            `json:"required,omitempty"`
-	Optional        bool            `json:"optional,omitempty"`
-	Computed        bool            `json:"computed,omitempty"`
-	Sensitive       bool            `json:"sensitive,omitempty"`
+	AttributeType       json.RawMessage `json:"type,omitempty"`
+	AttributeNestedType *nestedType     `json:"nested_type,omitempty"`
+	Description         string          `json:"description,omitempty"`
+	DescriptionKind     string          `json:"description_kind,omitempty"`
+	Deprecated          bool            `json:"deprecated,omitempty"`
+	Required            bool            `json:"required,omitempty"`
+	Optional            bool            `json:"optional,omitempty"`
+	Computed            bool            `json:"computed,omitempty"`
+	Sensitive           bool            `json:"sensitive,omitempty"`
+}
+
+type nestedType struct {
+	Attributes  map[string]*attribute `json:"attributes,omitempty"`
+	NestingMode string                `json:"nesting_mode,omitempty"`
+	MinItems    uint64                `json:"min_items,omitempty"`
+	MaxItems    uint64                `json:"max_items,omitempty"`
 }
 
 func marshalStringKind(sk configschema.StringKind) string {
@@ -27,12 +36,7 @@ func marshalStringKind(sk configschema.StringKind) string {
 }
 
 func marshalAttribute(attr *configschema.Attribute) *attribute {
-	// we're not concerned about errors because at this point the schema has
-	// already been checked and re-checked.
-	attrTy, _ := attr.Type.MarshalJSON()
-
-	return &attribute{
-		AttributeType:   attrTy,
+	ret := &attribute{
 		Description:     attr.Description,
 		DescriptionKind: marshalStringKind(attr.DescriptionKind),
 		Required:        attr.Required,
@@ -41,4 +45,27 @@ func marshalAttribute(attr *configschema.Attribute) *attribute {
 		Sensitive:       attr.Sensitive,
 		Deprecated:      attr.Deprecated,
 	}
+
+	// we're not concerned about errors because at this point the schema has
+	// already been checked and re-checked.
+	if attr.Type != cty.NilType {
+		attrTy, _ := attr.Type.MarshalJSON()
+		ret.AttributeType = attrTy
+	}
+
+	if attr.NestedType != nil {
+		nestedTy := nestedType{
+			MinItems:    uint64(attr.NestedType.MinItems),
+			MaxItems:    uint64(attr.NestedType.MaxItems),
+			NestingMode: nestingModeString(attr.NestedType.Nesting),
+		}
+		attrs := make(map[string]*attribute, len(attr.NestedType.Attributes))
+		for k, attr := range attr.NestedType.Attributes {
+			attrs[k] = marshalAttribute(attr)
+		}
+		nestedTy.Attributes = attrs
+		ret.AttributeNestedType = &nestedTy
+	}
+
+	return ret
 }

--- a/command/jsonprovider/block.go
+++ b/command/jsonprovider/block.go
@@ -24,24 +24,10 @@ func marshalBlockTypes(nestedBlock *configschema.NestedBlock) *blockType {
 		return &blockType{}
 	}
 	ret := &blockType{
-		Block:    marshalBlock(&nestedBlock.Block),
-		MinItems: uint64(nestedBlock.MinItems),
-		MaxItems: uint64(nestedBlock.MaxItems),
-	}
-
-	switch nestedBlock.Nesting {
-	case configschema.NestingSingle:
-		ret.NestingMode = "single"
-	case configschema.NestingGroup:
-		ret.NestingMode = "group"
-	case configschema.NestingList:
-		ret.NestingMode = "list"
-	case configschema.NestingSet:
-		ret.NestingMode = "set"
-	case configschema.NestingMap:
-		ret.NestingMode = "map"
-	default:
-		ret.NestingMode = "invalid"
+		Block:       marshalBlock(&nestedBlock.Block),
+		MinItems:    uint64(nestedBlock.MinItems),
+		MaxItems:    uint64(nestedBlock.MaxItems),
+		NestingMode: nestingModeString(nestedBlock.Nesting),
 	}
 	return ret
 }
@@ -74,4 +60,21 @@ func marshalBlock(configBlock *configschema.Block) *block {
 	}
 
 	return &ret
+}
+
+func nestingModeString(mode configschema.NestingMode) string {
+	switch mode {
+	case configschema.NestingSingle:
+		return "single"
+	case configschema.NestingGroup:
+		return "group"
+	case configschema.NestingList:
+		return "list"
+	case configschema.NestingSet:
+		return "set"
+	case configschema.NestingMap:
+		return "map"
+	default:
+		return "invalid"
+	}
 }

--- a/command/jsonprovider/provider_test.go
+++ b/command/jsonprovider/provider_test.go
@@ -51,6 +51,25 @@ func TestMarshalProvider(t *testing.T) {
 									Optional:        true,
 									DescriptionKind: "plain",
 								},
+								"volumes": {
+									AttributeNestedType: &nestedType{
+										NestingMode: "list",
+										Attributes: map[string]*attribute{
+											"size": {
+												AttributeType:   json.RawMessage(`"string"`),
+												Required:        true,
+												DescriptionKind: "plain",
+											},
+											"mount_point": {
+												AttributeType:   json.RawMessage(`"string"`),
+												Required:        true,
+												DescriptionKind: "plain",
+											},
+										},
+									},
+									Optional:        true,
+									DescriptionKind: "plain",
+								},
 							},
 							BlockTypes: map[string]*blockType{
 								"network_interface": {
@@ -141,6 +160,16 @@ func testProvider() *terraform.ProviderSchema {
 				Attributes: map[string]*configschema.Attribute{
 					"id":  {Type: cty.String, Optional: true, Computed: true},
 					"ami": {Type: cty.String, Optional: true},
+					"volumes": {
+						Optional: true,
+						NestedType: &configschema.Object{
+							Nesting: configschema.NestingList,
+							Attributes: map[string]*configschema.Attribute{
+								"size":        {Type: cty.String, Required: true},
+								"mount_point": {Type: cty.String, Required: true},
+							},
+						},
+					},
 				},
 				BlockTypes: map[string]*configschema.NestedBlock{
 					"network_interface": {

--- a/command/providers_schema_test.go
+++ b/command/providers_schema_test.go
@@ -9,7 +9,11 @@ import (
 	"testing"
 
 	"github.com/google/go-cmp/cmp"
+	"github.com/hashicorp/terraform/configs/configschema"
+	"github.com/hashicorp/terraform/providers"
+	"github.com/hashicorp/terraform/terraform"
 	"github.com/mitchellh/cli"
+	"github.com/zclconf/go-cty/cty"
 )
 
 func TestProvidersSchema_error(t *testing.T) {
@@ -28,8 +32,6 @@ func TestProvidersSchema_error(t *testing.T) {
 }
 
 func TestProvidersSchema_output(t *testing.T) {
-	// there's only one test at this time. This can be refactored to have
-	// multiple test cases in individual directories as needed.
 	fixtureDir := "testdata/providers-schema"
 	testDirs, err := ioutil.ReadDir(fixtureDir)
 	if err != nil {
@@ -48,11 +50,11 @@ func TestProvidersSchema_output(t *testing.T) {
 			defer testChdir(t, td)()
 
 			providerSource, close := newMockProviderSource(t, map[string][]string{
-				"test": []string{"1.2.3"},
+				"test": {"1.2.3"},
 			})
 			defer close()
 
-			p := showFixtureProvider()
+			p := providersSchemaFixtureProvider()
 			ui := new(cli.MockUi)
 			m := Meta{
 				testingOverrides: metaOverridesForProvider(p),
@@ -108,4 +110,46 @@ type providerSchema struct {
 	Provider          interface{}            `json:"provider,omitempty"`
 	ResourceSchemas   map[string]interface{} `json:"resource_schemas,omitempty"`
 	DataSourceSchemas map[string]interface{} `json:"data_source_schemas,omitempty"`
+}
+
+// testProvider returns a mock provider that is configured for basic
+// operation with the configuration in testdata/providers-schema.
+func providersSchemaFixtureProvider() *terraform.MockProvider {
+	p := testProvider()
+	p.GetProviderSchemaResponse = providersSchemaFixtureSchema()
+	return p
+}
+
+// providersSchemaFixtureSchema returns a schema suitable for processing the
+// configuration in testdata/providers-schema.ß
+func providersSchemaFixtureSchema() *providers.GetProviderSchemaResponse {
+	return &providers.GetProviderSchemaResponse{
+		Provider: providers.Schema{
+			Block: &configschema.Block{
+				Attributes: map[string]*configschema.Attribute{
+					"region": {Type: cty.String, Optional: true},
+				},
+			},
+		},
+		ResourceTypes: map[string]providers.Schema{
+			"test_instance": {
+				Block: &configschema.Block{
+					Attributes: map[string]*configschema.Attribute{
+						"id":  {Type: cty.String, Optional: true, Computed: true},
+						"ami": {Type: cty.String, Optional: true},
+						"volumes": {
+							NestedType: &configschema.Object{
+								Nesting: configschema.NestingList,
+								Attributes: map[string]*configschema.Attribute{
+									"size":        {Type: cty.String, Required: true},
+									"mount_point": {Type: cty.String, Required: true},
+								},
+							},
+							Optional: true,
+						},
+					},
+				},
+			},
+		},
+	}
 }

--- a/command/testdata/providers-schema/basic/output.json
+++ b/command/testdata/providers-schema/basic/output.json
@@ -30,6 +30,25 @@
                                 "optional": true,
                                 "computed": true,
                                 "description_kind": "plain"
+                            },
+                            "volumes": {
+                                "nested_type": {
+                                    "nesting_mode": "list",
+                                    "attributes": {
+                                        "size": {
+                                            "type": "string",
+                                            "required": true,
+                                            "description_kind": "plain"
+                                        },
+                                        "mount_point": {
+                                            "type": "string",
+                                            "required": true,
+                                            "description_kind": "plain"
+                                        }
+                                    }
+                                },
+                                "description_kind": "plain",
+                                "optional": true
                             }
                         },
                         "description_kind": "plain"

--- a/command/testdata/providers-schema/required/output.json
+++ b/command/testdata/providers-schema/required/output.json
@@ -30,6 +30,25 @@
                                 "optional": true,
                                 "computed": true,
                                 "description_kind": "plain"
+                            },
+                            "volumes": {
+                                "nested_type": {
+                                    "nesting_mode": "list",
+                                    "attributes": {
+                                        "size": {
+                                            "type": "string",
+                                            "required": true,
+                                            "description_kind": "plain"
+                                        },
+                                        "mount_point": {
+                                            "type": "string",
+                                            "required": true,
+                                            "description_kind": "plain"
+                                        }
+                                    }
+                                },
+                                "description_kind": "plain",
+                                "optional": true
                             }
                         },
                         "description_kind": "plain"

